src/ServiceControl.Monitoring/Metrics/Raw/LongValueOccurrences.cs

```diff
@@ -1,23 +1,63 @@
 ﻿namespace ServiceControl.Monitoring.Metrics.Raw
 {
+    using System;
+    using System.Collections.Concurrent;
     using NServiceBus;
 
     public class LongValueOccurrences : IMessage
     {
-        public long Version;
+        public class Pool
+        {
+            public static readonly Pool Default = new Pool();
+            ConcurrentStack<LongValueOccurrences> pool = new ConcurrentStack<LongValueOccurrences>();
 
-        public long BaseTicks;
+            public LongValueOccurrences Lease()
+            {
+                LongValueOccurrences value;
+                if (pool.TryPop(out value))
+                {
+                    return value;
+                }
 
-        public int[] Ticks;
+                return new LongValueOccurrences();
+            }
 
-        public long[] Values;
-    }
+            public void Release(LongValueOccurrences message)
+            {
+                message.index = InitialIndex;
+                Array.Clear(message.entries, 0, MaxEntries);
+                pool.Push(message);
+            }
+        }
 
-    public class ProcessingTime : LongValueOccurrences
-    {
-    }
+        int index = InitialIndex;
+        public int Length => index;
+        public Entry[] entries = new Entry[MaxEntries];
+        const int MaxEntries = 512;
+        const int InitialIndex = 0;
 
-    public class CriticalTime : LongValueOccurrences
-    {
+        public bool TryRecord(long dateTicks, long value)
+        {
+            if (index == MaxEntries)
+            {
+                return false;
+            }
+
+            entries[index].DateTicks = dateTicks;
+            entries[index].Value = value;
+
+            index += 1;
+
+            return true;
+        }
+
+        public long MinDateTick => entries[0].DateTicks;
+        public long MaxDateTick => entries[index].DateTicks;
```

Comment:
That's the format currently used by `NSB.Metrics`. Values are sorted in a single message before sending them out to `SC.Monitoring`. I don't know if there's a test for it, but I can introduce one if there's none to make it a part of the contract.